EIPS/eip-4804.md

```diff
@@ -0,0 +1,158 @@
+
+#### Example 2
+
+```
+web3://0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b:4
+```
+
+The protocol will call the address with "From" = "0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b" and "Calldata" = "0x" with chainid = 4.
+
+#### Example 3
+
+```
+web3://cyberbrokers-meta.eth:1/renderBroker/9999
+```
+
+The protocol will find the address of **cyberbrokers-meta.eth** from ENS, and then call the address with "From" = "0x..." and "Calldata" = "0x" + `keccak("view(uint256)")[0:4] + abi.encode(uint256(9999))` with chainid = 1.
+
+#### Exmaple 4
+
+```
+web3://home.w3q:3334/
+```
+
+The protocol will find the address of **home.w3q** from W3NS, and then call the address with "From" = "0x..." and "Calldata" = "0x" with chainid = 3334.
+
+#### Example 5
+
+```
+web3://wusdt.eth:4->(uint256)/balanceOf/charles.eth
+```
+
+The protocol will find the addresses of **wusdt.eth** and **charles.eth** and then call the method "balanceOf(address)" of the contract with the **charles.eth**'s address. The returned data will be parsed as uint256 like `[ "10000000000000" ]`.
+
+#### Example 6
+```
+web3://wusdt.eth:4->()/balanceOf/charles.eth
+```
+
+The protocol will find the address of **wusdt.eth** and then call the method "balanceOf(address)" of the address. The returned data will be parsed as raw bytes like `["0x000000000000000000000000000000000000000000000000000009184e72a000"]`.
+
+## Rationale
+
+The purpose of the proposal is to add a decentralized presentation layer for Ethereum.  With the layer, we are able to render any web content (including HTML/CSS/JPG/PNG/SVG, etc) on-chain using human-readable URLs, and thus EVM can be served as decentralized Backend.  The design of the standard is based on the following principles:
+
+- **Human-readable**.  The Web3 URL should be easily recognized by human similar to Web2 URL (http://).  As a result, we support names from name services to replace address for better readability.  In addition, instead of using calldata in hex, we use human-readable method + arguments and translate them to calldata for better readability.
+
+- **Maximum-Compatible with HTTP-URL standard**.  The Web3 URL should be compatible with HTTP-URL standard including relative pathing, query, fragment, etc so that the support of existing HTTP-URL (e.g., by browser) can be easily extended to Web3 URL with minimal modification.  This also means that existing Web2 users can easily migrate to Web3 with minimal extra knowledge of this standard.
+
+- **Simple**.  Instead of providing explict types in arguments, we use a "maximum likelihood" principle of auto-detecting the types of the arguments such as address, bytes32, and uint256.  This could greatly minimize the length of URL, while avoiding confusion.  In addition, explict types are also supported to clear the confusion if necessary.
```

Comment:
```suggestion
- **Simple**.  Instead of providing explicit types in arguments, we use a "maximum likelihood" principle of auto-detecting the types of the arguments such as address, bytes32, and uint256.  This could greatly minimize the length of URL, while avoiding confusion.  In addition, explicit types are also supported to clear the confusion if necessary.
```